### PR TITLE
Add size based roll

### PR DESCRIPTION
**Pull request by PhoenixRion**
Missing test to verify max file size value

### CODE REVIEW

File: src/test/java/com/arpnetworking/metrics/impl/BaseFileSinkTest.java [hunk 1]
```diff
@@ -21,7 +21,7 @@
 import ch.qos.logback.classic.spi.ILoggingEvent;
 import ch.qos.logback.core.encoder.Encoder;
 import ch.qos.logback.core.rolling.RollingFileAppender;
-import ch.qos.logback.core.rolling.TimeBasedRollingPolicy;
+import ch.qos.logback.core.rolling.SizeAndTimeBasedRollingPolicy;
```
Comment (vjkoskela) [inline src/test/java/com/arpnetworking/metrics/impl/BaseFileSinkTest.java]: This version of logback does not correctly support the maximum number of files to rotate when size based rotation is enabled. You can either ugprade Logback and verify the desired behavior or we have a custom policy in logback-steno (which is already a dependency) which correctly supports size based rotation with random time based rotation as well:

https://github.com/ArpNetworking/logback-steno/blob/master/src/main/java/com/arpnetworking/logback/SizeAndRandomizedTimeBasedFNATP.java

File: src/main/java/com/arpnetworking/metrics/impl/BaseFileSink.java [hunk 2]
```diff
@@ -51,21 +51,23 @@ protected Logger getMetricsLogger() {
...
 
-    private TimeBasedRollingPolicy<ILoggingEvent> createRollingPolicy(
+    private SizeAndTimeBasedRollingPolicy<ILoggingEvent> createRollingPolicy(
             final String extension,
             final String fileNameWithoutExtension,
             final int maxHistory,
+            final String maxFileSize,
             final boolean compress) {
 
         // TODO(vkoskela): Use LogbackSteno's random compression strategy [MAI-413]
```
Comment (vjkoskela) [inline src/main/java/com/arpnetworking/metrics/impl/BaseFileSink.java]: See my note below, we could solve this at the same time.

File: src/test/java/com/arpnetworking/metrics/impl/BaseFileSinkTest.java [hunk 3]
```diff
@@ -129,7 +137,10 @@ public void testBuilderWithNull() throws IOException {
         Assert.assertEquals(24, rollingPolicy.getMaxHistory());
         Assert.assertTrue(encoder.isImmediateFlush());
         Assert.assertEquals(expectedPath + "query.log", rollingAppender.getFile());
-        Assert.assertEquals(expectedPath + "query.%d{yyyy-MM-dd-HH}.log.gz", rollingPolicy.getFileNamePattern());
+        Assert.assertEquals(expectedPath + "query.%d{yyyy-MM-dd-HH}.%i.log.gz", rollingPolicy.getFileNamePattern());
+
+        // Not currently implemented by logback
+//        Assert.assertEquals("100MB", rollingPolicy.getMaxFileSize());
```
Comment (vjkoskela) [inline src/test/java/com/arpnetworking/metrics/impl/BaseFileSinkTest.java]: Given the issues with size based rolling I think it would be prudent to include an integration test. For example, set a very low threshold so each even is written to its own log file. You can have a few simple tests and then a more complex test where you exceed the maximum number of files based on size only and that the oldest file(s) is deleted.

**Comment by PhoenixRion** [comment]
@vjkoskela This line is a problem for a unit test that checks for the deletion from totalSizeCap as it would require the test to run for at least two hours.  I checked the current master and this has been removed, but no idea when the next release will go out.  https://github.com/qos-ch/logback/blob/2dad0019ebaff91a183200dab54e1e4a94407d5a/logback-core/src/main/java/ch/qos/logback/core/rolling/helper/TimeBasedArchiveRemover.java#L107

**Comment by vjkoskela** [comment]
@PhoenixRion so this just means there will be a minimum of two files regardless of size. if you set the rotation to 1kb, max files to 3, then compute and set total cap to 3x 1kb. Now write 4x 1kb and you should have only three files of 1kb each with content 2, 3, 4 (content 1 was deleted as the oldest). I don't understand what this has to do with time based rotation.

File: src/main/java/com/arpnetworking/metrics/impl/BaseFileSink.java [hunk 4]
```diff
@@ -414,5 +453,7 @@ protected void validate(final List<String> failures) {
         private static final Boolean DEFAULT_PRUDENT = Boolean.FALSE;
         private static final Boolean DEFAULT_DROP_WHEN_QUEUE_FULL = Boolean.FALSE;
         private static final Integer DEFAULT_MAX_QUEUE_SIZE = Integer.valueOf(500);
+        private static final String DEFAULT_MAX_FILE_SIZE = "100MB";
+        private static final String DEFAULT_TOTAL_SIZE_CAP = "20GB";
```
Comment (vjkoskela) [inline src/main/java/com/arpnetworking/metrics/impl/BaseFileSink.java]: The total cap size should be set to maximum files times max file size.

**Comment by PhoenixRion** [comment]
@vjkoskela The offset value that is being compared against is the time based rolling period, not the count of files.  The max history parameter is used in the same fashion.

File: src/main/java/com/arpnetworking/metrics/impl/BaseFileSink.java (hunk 4 above)
Comment (PhoenixRion) [inline src/main/java/com/arpnetworking/metrics/impl/BaseFileSink.java]: Max history is only used against the time based rolling period, as far as I have seen.

Comment (vjkoskela) [inline src/main/java/com/arpnetworking/metrics/impl/BaseFileSink.java]: Yes, this is a bug in the client configuration. Max history should be the number of files to keep in my opinion. @BrandonArp ?

That means across time and space rolling policies. For example, without a size constraint it should be the number hours. However, with a size constraint it may be fewer than that many hours. 

These configuration options on the client are not intended to duplicate the Logback configuration but to stand on their own.

I suggest:

 - No default maximum size.
 - Setting maximum size sets totalSizeCap on logback to maxHistory times maximum size.
 - Tests for enforcement of max history based on size (since this has been a buggy feature in logback).

**Comment by vjkoskela** [comment]
@PhoenixRion I'm not sure that I follow. Please see my comment above, if we are still not on the same page please reach out over chat and we can discuss further.

File: src/main/java/com/arpnetworking/metrics/impl/BaseFileSink.java (hunk 4 above)
Comment (PhoenixRion) [inline src/main/java/com/arpnetworking/metrics/impl/BaseFileSink.java]: No default max size can easily return a time based rolling policy only.

totalSizeCap = maxHistory * maxFileSize  only makes sense when the compression is set to false.

Testing enforcement of max size is difficult in the current release of logback due to code in the archive remover that guards the most recent two periods of time (depending on your time based rolling schedule).  This code has been removed in the current master and I was able to write a good test.

Comment (vjkoskela) [inline src/main/java/com/arpnetworking/metrics/impl/BaseFileSink.java]: I'm not sure that I agree with "only makes sense when the compression is set to false.". Can you elaborate?

Or was this cleared up during our conversation today?

Comment (vjkoskela) [inline src/main/java/com/arpnetworking/metrics/impl/BaseFileSink.java]: So this may be a difference between how Logback and LogackSteno implemented it. I would have to check but my understanding was that max history limits the number of files not the number of periods. Then max size limits the size of each file. 

Can you confirm which behavior Logback and Steno each use? If it is as you say that max history only applies to time periods then we do indeed have a problem; that max size cannot limit total size effectively on its own.

Comment (PhoenixRion) [inline src/main/java/com/arpnetworking/metrics/impl/BaseFileSink.java]: The relevant logic lives in the ArchiveRemover which is common to both rolling policies.  MaxSize will limit total size under the current logback release as long as you don't hit that limit before 2 rolling periods.  (2 hours for most of our services)  This does make writing a unit test a bit of a pain.

Max history does equal the number of files when doing only time based rolling but that does not currently remain true when you start rolling by size.

**Comment by PhoenixRion** [comment]
Due to orginization changes this PR will have to be continued on a new PR from a fork.  https://github.com/ArpNetworking/metrics-client-java/pull/27
